### ISSUE TITLE
Parallel processing with map and friends

### ISSUE THREAD

**Issue by lmullen** [hadley/purrr#194]
Would it be possible in a future version of purrr to have a multicore version of `map()`? E.g., `mcmap` is to `map` as `mclapply` is to `apply`? Perhaps this is out of scope for the package.

**Comment by hadley** [comment]
See #121

**Comment by sdanielzafar** [comment]
I thought `pmap()` covered this?

**Comment by hadley** [comment]
@sdanielzafar different definition of parallel

**Comment by csrvermaak** [comment]
Using the `future' package, you can try `map(Var1, ~value(future(yourfunction(.x))))`

**Comment by sdanielzafar** [comment]
The syntactic sugar is just not there, I typically just use `pblapply(..., cl = cl)` to parellelize. Plus you get nice progress bars. Is there an advantage to this `future` solution over `pblapply`?

**Comment by dangulod** [comment]
Hi @csrvermaak does it really works? I tried using future and map function together as you mentioned, but nothing happens.

**Comment by csrvermaak** [comment]
@dangulod - what do you mean "nothing happens"?

**Comment by dangulod** [comment]
I am trying to process in parallel the following function:

````
conv_pois_norm = function(l = l, mean = mean, sd = sd) { 

  return(sum(rnorm(n = rpois(n = 1, lambda = l), mean = mean, sd = sd)))

}
````
As I understood in your comment I can do it with the future package, so I try with:

````
library(purrr)
library(future)

n_simul = 1e6
mean = 10
sd = 20
l = 10

conv_pois_norm(l, mean, sd)

x = rep(0, n_simul)

plan(multiprocess)

x = map(x, ~future(conv_pois_norm(l, mean, sd)))
````

When i execute this function, my computer start working and never stops.

Probably, this does not work on Windows, or I understood something wrong. Thank you for your help

**Comment by dangulod** [comment]
@csrvermaak - I am sorry. After several tries, I finally understand how it works. The real problem was the high number of simulations to perform and the multi core plan is not currently supported on windows.

**Comment by csrvermaak** [comment]
Firstly, you need to convert x to a `tibble` and use `mutate(map())`.

Secondly, you are spawning 1e6 seperate R processes, one for each `n_simul`, thus it would take ages. There is some overhead to creating an R process, and I don't think you're using the future correctly here.

Thirdly, you're not extracting the result using `value()`

I tend to use it for long slow functions, running about 20/30 in parallel, not 1e6.



HTH

**Comment by csrvermaak** [comment]
I've tried to update your example to prove the concept. I added x as an argument to your function, so you can see how to use it.

```
conv_pois_norm = function(x, l = l, mean = mean, sd = sd) { 
  
  return(x + sum(rnorm(n = rpois(n = 1, lambda = l), mean = mean, sd = sd)))
  
}


library(purrr)
library(future)
library(tibble)
library(dplyr)

n_simul = 20
mean = 10
sd = 20
l = 10

conv_pois_norm(10, l, mean, sd)

x <-  tibble(x = rep(0, n_simul))

plan(multiprocess)

x <- x %>% 
  mutate(myFuture = map(x, ~future(conv_pois_norm(.x, l, mean, sd)))) %>% 
  mutate(myResult = map_dbl(myFuture, ~value(.x)))   # Wait for Resolved

x
```

**Comment by dangulod** [comment]
Thank for you for your reply. Now, I understand how it really works

**Comment by RenBinary** [comment]
Any idea on how could this be used with pmap?

pmap(list(1:10,1:10,1:10),~future(..1+..2+..3)) %>% map(value) throws errors that it doesn't know what ..1 and the others are
